connectivity/tests/bgp.go

```diff
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package tests
+
+import (
+	"context"
+	"fmt"
+
+	ciliumv2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	slimv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/apis/meta/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+
+	"github.com/cilium/cilium-cli/connectivity/check"
+	"github.com/cilium/cilium-cli/utils/features"
+)
+
+const (
+	bgpPeeringPolicyName = "test-bgp-peering-policy"
+	bgpAdvertisementName = "test-bgp-advertisement"
+	bgpPeerConfigName    = "test-bgp-peer-config"
+	bgpClusterConfigName = "test-bgp-cluster-config"
+
+	bgpCiliumASN = 65001
+	bgpFRRASN    = 65000
+
+	bgpCommunityPodCIDR = "65001:100"
+	bgpCommunityService = "65001:200"
+)
+
+func BGPAdvertisements(bgpAPIVersion uint8) check.Scenario {
+	return &bgpAdvertisements{
+		bgpAPIVersion: bgpAPIVersion,
+	}
+}
+
+type bgpAdvertisements struct {
+	bgpAPIVersion uint8
+}
+
+func (s *bgpAdvertisements) Name() string {
+	return fmt.Sprintf("bgpv%d-advertisements", s.bgpAPIVersion)
+}
+
+func (s *bgpAdvertisements) Run(ctx context.Context, t *check.Test) {
+	ct := t.Context()
+
+	t.ForEachIPFamily(func(ipFamily features.IPFamily) {
+		defer func() {
+			s.cleanup(ctx, t)
+		}()
+
+		// configure FRR
+		frrPeers := ct.InternalNodeIPAddresses(ipFamily)
+		frrConfig := check.RenderFRRBGPPeeringConfig(t, frrPeers)
+		for _, frr := range ct.FRRPods() {
+			check.ApplyFRRConfig(ctx, t, &frr, frrConfig)
+		}
+
+		// configure BGP on Cilium
+		if s.bgpAPIVersion == 1 {
+			s.configureBGPv1Peering(ctx, t, ipFamily)
+		} else {
+			s.configureBGPv2Peering(ctx, t, ipFamily)
+		}
+
+		// wait for BGP peers and expected prefixes
+		podCIDRPrefixes := ct.PodCIDRPrefixes(ipFamily)
+		svcPrefixes := ct.EchoServicePrefixes(ipFamily)
+		for _, frr := range ct.FRRPods() {
+			check.WaitForFRRBGPNeighborsState(ctx, t, &frr, frrPeers, "Established")
+
+			frrPrefixes := check.WaitForFRRBGPPrefixes(ctx, t, &frr, podCIDRPrefixes, ipFamily)
+			check.AssertFRRBGPCommunity(t, frrPrefixes, podCIDRPrefixes, bgpCommunityPodCIDR)
+
+			frrPrefixes = check.WaitForFRRBGPPrefixes(ctx, t, &frr, svcPrefixes, ipFamily)
+			if s.bgpAPIVersion != 1 { // BGPv1 does not support path attributes for ClusterIP service advertisements
+				check.AssertFRRBGPCommunity(t, frrPrefixes, svcPrefixes, bgpCommunityService)
+			}
+		}
+
+		i := 0
```

Comment:
I am bit confused by usage of `i` here. Looks like it monotonically increase from pods to services. 

`curl-echo-pod-v4-<0..n>`
`curl-echo-service-v4-<n+1...>` any why not do `curl-echo-service-v4-<0..n>`

Maybe add a comment explaining why it is this way.

Reply:
It was a copy-paste error, fixed.